include/elements/2d/quadrilateral_bspline_element.h

```diff
@@ -0,0 +1,245 @@
+#ifndef MPM_BSPLINE_ELEMENT_H_
+#define MPM_BSPLINE_ELEMENT_H_
+
+#include "quadrilateral_element.h"
+
+namespace mpm {
+
+//! Quadrilateral BSpline element class derived from Quadrilateral
+//! \brief Quadrilateral BSpline element
+//! \details quadrilateral BSpline element with adaptive number of nodes \n
+//! Type of elements and sizes of shape function, gradient shape function,
+//! B-matrix, indices are according to the number of nodes \n <pre>
+//!
+//! Type 1: Regular Element: nnodes = 16
+//!
+//!   13----------12----------11----------10
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   14----------3-----------2-----------9
+//!   |           |           |           |
+//!   |           | particle  |           |
+//!   |           | location  |           |
+//!   |           |           |           |
+//!   15----------0-----------1-----------8
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   4-----------5-----------6-----------7
+//!
+//! Type 2: Boundary Element: nnodes < 16 (cell located at the mesh boundary)
+//! e.g. Mesh edge
+//!
+//!   9-----------8-----------7-----------6
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   |           |           |           |
+//!   10----------3-----------2-----------5
+//!   |           |           |           |
+//!   |           | particle  |           |
+//!   |           | location  |           |
+//!   |           |           |           |
+//!   11----------0-----------1-----------4
+//!
+//! e.g. Mesh corner
+//!
+//!   8-----------7-----------6
+//!   |           |           |
+//!   |           |           |
+//!   |           |           |
+//!   |           |           |
+//!   3-----------2-----------5
+//!   |           |           |
+//!   | particle  |           |
+//!   | location  |           |
+//!   |           |           |
+//!   0-----------1-----------4
```

Comment:
so "edge" and "corner" elements are both Type 2? I.e., anything with <16 nnodes is treated the same?

Reply:
In fact, all the treatments are the same! Type 1 and 2 are just to explain that there is a possibility to have fewer number of nodes per element. Maximum will be 16 in 2D and 64 in 3D.